rfcs/detecting_sensitive_textual_content/20230309-project_proposal_detecting_sensitive_textual_content.md

```diff
@@ -0,0 +1,379 @@
+hash and an API library to produce the hash from the thumbnail.
+
+It has the following downsides:
+
+- Significantly more complicated than Gaussian blur
+- Requires conditionally calculating the hash on the API side in the thumbnail
+  endpoint
+  - This would be done by using a new query parameter on the thumbnail endpoint
+    to request the blurred hash
+- Can have a performance impact on the thumbnail endpoint
+- Introduces a delay between when the image is unblurred when it is displayed
+  due to needing to download the actual thumbnail
+- Could have different performance issues if decoding the hash is a heavy
+  operation for low-spec clients
+  - Looking at the code, I highly doubt this because it's a
+    [fast bit of maths](https://github.com/mad-gooze/fast-blurhash/blob/main/index.js)
+    that (due to the nature of BlurHash) ignores almost all the content of the
+    image rather than a Gaussian blur which needs to consider ever aspect of the
+    image in order to apply the blur
+
+It has the following upsides:
+
+- Does not require the client to download the image just to blur it
+  - If the client never unblurs an image, it will have saved all the bytes of
+    the image from being downloaded to the client
+- We can easily cache the hash in Redis to ameliorate performance over time
+  - If we do so, then we could use the hash for other things down the line (like
+    average colour filters) and add it directly to the documents via the same
+    ETL pipeline we'll develop for dead links, further ameliorating any API
+    performance issues
+- Does not rely on the client's ability to apply and render a Gaussian blur
+  transformation to potentially several images on a single page
+- Extremely minor and only barely relevant: If we calculated the BlurHash for
+  all images in the future in our re-crawler, we could use them as placeholders
+  rather than the skeleton boxes we currently use
+
+#### Recommendation
+
+If we can spare the time, I think BlurHash is the correct choice for the
+following reasons:
+
+1. While significantly more complex, it is not sufficiently complex to be
+   burdensome. The simplicity of the CSS filter makes BlurHash look monstrous,
+   but in the grand scheme of things, it's actually a relatively simple approach
+   to blurring the images.
+2. It avoids performance issues with relying on low-spec clients to apply
+   Gaussian blurs to potentially several images.
+3. BlurHash hashes have significant future value and all potential API level
+   performance issues can be ameliorated through caching hashes with Redis.
+
+The only significant downside that I think exists is the delay between
+unblurring and displaying the full image due to needing to download the
+full unblurred image. However, I think this is balanced out by the bytes saved
+over the network in _not_ sending images that might potentially never be
+unblurred.
+
+For reviewers: **What are you feelings on this issue and why?**. One alternative
```

Comment:
I 100% agree the future benefits of blurhash and love providing the hashes themselves _as a feature_ to our users. It could be _extremely_ useful for API consumers and also dramatically increases the safety of Openverse _integrations_ using the API. So I am fully convinced blurhash does _more_ to contribute to safety in Openverse as a project beyond the frontend.

That said, the blur CSS is _so trivial_ to implement and allows for rapid, iterative adjustment based on user feedback. If we hear that lots of users think it's _too_ blurry or insufficiently blurry, we could adapt it. Because it's literally a one-line thing to implement, I would advocate for **initializing this project with CSS blurring** and creating a new content safety project to implement blurhash in the future. 

This is just one opinion, and I fully admit that it's a painful one for me to make. but I do also fundamentally think separating these things would allow us to ship meaningful safety improvements faster, without limiting our ability to implement blurhash later or having wasted time on a temporary solution (since the CSS is so trivial).

Reply:
I agree with this. I'll change the proposal to make this the recommendation. That allows us to push off deciding between BlurHash or some other solution like LQIP.